[Puri Virakarin] iP

purivirakarin opened the pull request:
### AtsisBot

> We can help you manage all your todolist :)

AtsisBot is an essential tool for you to help you with

- text-based interaction
- friendly user experience
- fast and efficient

All you have to do is

1. download it from [here](https://github.com/purivirakarin/ip/releases/download/v0.1/duke.jar).
2. run it
3. add your todos/deadlines/events
4. let it manage your task for you 😉

Features:

- [x]  Managing tasks
- [x]  Managing deadlines
- [x]  Managing events
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```
public static void main(String[] args) {
    storage = new Storage("data/tasks.txt");
    ...
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Implement an initial skeletal

Commit:
features: add the echo command

Commit:
Add, List

Commit:
Mark as Done

Commit:
ToDos, Events, Deadlines

Commit:
Automated Text UI Testing

Commit:
Handle Errors

Commit:
Delete

Commit:
Fix the git track error, add task and taskstatus file, use enums, add javadocs

Commit:
Add the load and unload tasks from file

Commit:
Merge branch 'branch-Level-7'

Commit:
Add date and time format to deadline and event

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor tasks & Add Parser Storage Ui Command CommandEnum classes

Commit:
Merge branch 'add-more-oop'

Commit:
Refactor atsisbot package by moving under atsisbot folder

Commit:
Merge branch 'Refactor-atsisbot-package'

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

Commit:
Add JUnit tests for Parser and Storage

Commit:
Merge branch 'add-junit-test'

Commit:
Add jar file support in gradle.buile

Commit:
Merge branch 'add-jar-file'

Commit:
Add missing Javadocs to classes and methods

Commit:
Edit coding standard to follow se-ed conventions

https://se-education.org/guides/conventions/java/intermediate.html

Commit:
Add find feature to search tasks

Commit:
Merge branch 'add-java-docs'

Commit:
Merge branch 'edit-coding-standard'

Commit:
Update style conflict from merge commit

Commit:
Merge branch 'add-find-feature'

KohGuanZeh commented on src/main/java/atsisbot/Command.java:
```diff
@@ -0,0 +1,163 @@
+package atsisbot;
+
+import atsisbot.task.*;
```
I think imported classes should always be listed explicitly?

KohGuanZeh commented on src/main/java/atsisbot/Command.java:
```diff
@@ -0,0 +1,163 @@
+     * @param args     The index of the task to be deleted.
+     * @param taskList The task list from which the task will be deleted.
+     */
+    public static void delete(String args, TaskList taskList) {
```
Maybe `args` can be renamed as it is a single `String` item?

KohGuanZeh commented on src/main/java/atsisbot/Storage.java:
```diff
@@ -0,0 +1,117 @@
+package atsisbot;
+
+import atsisbot.task.*;
+
+import java.io.*;
```
I think that imported classes should always be listed explicitly?

KohGuanZeh commented on src/main/java/atsisbot/Storage.java:
```diff
@@ -0,0 +1,117 @@
+    private Task readLine(String line) {
+        String[] taskInfo = line.split(" \\| ");
+        Task task = null;
+        switch (taskInfo[0]) {
+            case "T":
+                task = new Todo(taskInfo[2]);
+                break;
+            case "D":
+                try {
+                    LocalDateTime deadline = LocalDateTime.parse(taskInfo[3], Task.formatter);
+                    task = new Deadline(taskInfo[2], deadline);
+                } catch (DateTimeParseException e) {
+                    System.out.println(
+                            "Invalid deadline format in the atsisbot.task list: " + e.getMessage());
+                }
+                break;
+            case "E":
+                try {
+                    LocalDateTime startDateTime = LocalDateTime.parse(taskInfo[3], Task.formatter);
+                    LocalDateTime endDateTime = LocalDateTime.parse(taskInfo[4], Task.formatter);
+                    task = new Event(taskInfo[2], startDateTime, endDateTime);
+                } catch (DateTimeParseException e) {
+                    System.out.println("Invalid event format in the atsisbot.task list: " + e.getMessage());
+                }
+                break;
+            default:
+                break;
```
I think `case` is meant to be aligned with `switch` statement?

KohGuanZeh commented on src/main/java/atsisbot/Ui.java:
```diff
@@ -0,0 +1,174 @@
+ * displaying task lists.
+ */
+public class Ui {
+    private static final String line = "____________________________________________________________\n";
+    private static final String welcomeMessage = line + "Hello! I'm atsisbot.AtsisBot\n" + "What can I do for you?\n"
+            + line;
+    private static final String endingMessage = line + "Bye. Hope to see you again soon!\n" + line;
+    private static final String listMessage = "Here are the tasks in your list:\n";
+    private static final String markMessage = "Nice! I've marked this atsisbot.task as done:\n";
+    private static final String unmarkMessage = "OK, I've marked this atsisbot.task as not done yet:\n";
+    private static final String addTaskMessage = "Got it. I've added this atsisbot.task:\n";
+    private static final String deleteTaskMessage = "Noted. I've removed this atsisbot.task:\n";
+    private static final String noDescriptionMessage = "The description of a %s cannot be empty.\n";
+    private static final String invalidDeadlineFormatMessage = "Invalid deadline format. Please use: deadline <description> /by <date>\n";
+    private static final String invalidEventFormatMessage = "Invalid event format. Please use: event <description> /from <date> /to <date>\n";
+    private static final String invalidTaskNumberMessage = "Invalid atsisbot.task number. Please enter a valid atsisbot.task number.\n";
+    private static final String invalidInputFormatMessage = "Invalid input format. Please use the correct format.\n";
+    private static final String unknownCommandMessage = "I'm sorry, but I don't understand that command.";
+    private static final String invalidFindMessage = "No matching tasks found.\n";
+    private static final String findMessage = "Here are the matching tasks in your list:\n";
```
Maybe constants should be in screaming snake case?

ejnan replied:
I agree with this! According to our coding standard, **constant names must be all uppercase using underscore to separate words (aka SCREAMING_SNAKE_CASE)**

so maybe your code could look like this:
`private static final String WELCOME_MESSAGE = line + "Hello! I'm atsisbot.AtsisBot\n" + "What can I do for you?\n" + line;`

KohGuanZeh reviewed (commented):
I like that you have made every method and variable as intuitive as possible and liked that you have added JavaDocs for most, if not all important functions. I think there are some coding standards conventions you have missed from the guide such as the `switch` `case` alignment as well as explicit import of classes. Apart from that, LGTM.👍

ejnan commented on src/main/java/atsisbot/AtsisBot.java:
```diff
@@ -0,0 +1,86 @@
+   */
+  private static void response(CommandEnum commandEnum, String args) {
+    try {
+      switch (commandEnum) {
+        case LIST:
+          Command.list(taskList);
+          break;
+        case MARK:
+          Command.mark(args, taskList);
+          break;
+        case UNMARK:
+          Command.unmark(args, taskList);
+          break;
+        case DELETE:
+          Command.delete(args, taskList);
+          break;
+        case TODO:
+          Command.todo(args, taskList);
+          break;
+        case DEADLINE:
+          Command.deadline(args, taskList);
+          break;
+        case EVENT:
+          Command.event(args, taskList);
+          break;
+        case FIND:
+          Command.find(args, taskList);
+          break;
+        default:
+          Ui.printUnknownCommandMessage();
```
According to our Java coding standard, there should be no additional indentation for case clauses within a switch statement. Could you adjust the indentation level for the case clauses to match the standard?

```
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}
```

ejnan reviewed (commented):
Overall, I feel that your code is easy to read and well-structured. I like how your JavaDoc comments are clear, and naming conventions seem to follow the coding standard. 
I noted a few minor coding standard violations which were pointed out by the other reviewer before me, except for this one on switch statements,

Commit:
Add CheckStyle.xml

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add GUI using JavaFX FXML

Commit:
Merge branch 'branch-Level-10'

Commit:
Fix styling

Commit:
Edit the variable names for constants

Commit:
Add assertions to methods in AtsisBot and MainWindow classes

There are several methods that assume that certain conditions are true
and that the program will not crash if they are not. This is not a good.

This commit adds assertions to these methods to make sure that the
conditions are true and that the program will not crash if they are not.

This is being done by adding the assert keyword to the methods by
checking the conditions and throwing an AssertionError if the conditions
are not true.

purivirakarin committed:
Merge pull request #1 from purivirakarin/branch-A-Assertions

Add assertions to methods in AtsisBot and MainWindow classes

Commit:
Improve code quality by avoiding deep nesting

purivirakarin committed:
Merge pull request #2 from purivirakarin/branch-A-CodeQuality

Improve code quality by avoiding deep nesting

Commit:
Replace while loop into stream in Storage.java

Commit:
Merge branch 'branch-A-Streams'

Commit:
Add CI in .github/workflows/gradle.yml

Commit:
Add C-Priority extension

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Add Ui.png to document user interface layout

A new screenshot called Ui.png has been added to the `docs` directory to provide a visual representation of the current user interface layout. This image will assist developers and new contributors in understanding the UI components and their arrangement.

- Move the Ui.png screenshot to the `docs` folder
- Update documentation references to include the new UI layout image

Commit:
Merge branch 'add-ui-png' into current-branch

This merge integrates the 'add-ui-png' branch, which includes the updated Ui.png file in the `docs` directory. The addition of this UI screenshot is necessary for providing a clear, visual representation of the application's user interface, which is beneficial for both current contributors and future onboarding.

- Ensure the latest UI design is documented
- Provide a visual guide for new developers

Commit:
Store ATSISBot user guide in docs

The comprehensive user guide for ATSISBot has been added to the `docs` directory as `README.md`. This document will serve as a quick reference for new users and as a detailed guide for existing users to utilize ATSISBot features effectively.

- Place the ATSISBot user guide in the `docs` directory
- Rename the guide to `README.md` for easy access and visibility

Commit:
Merge branch 'branch-A-UserGuide' into master

This merge brings the latest updates from 'branch-A-UserGuide' into the master branch, including the new ATSISBot User Guide. The guide is pivotal for helping users understand and navigate through the functionalities of ATSISBot.

- Integrate comprehensive user guide updates into the main documentation
- Ensure all recent user guide enhancements are reflected in the master branch

Commit:
Fix StorageTest.java to account for priority output

The unit test in `src/test/java/atsisbot/StorageTest.java` has been updated to expect the correct output that now includes task priority levels (low/medium/high). This change aligns the test with the recent feature increment that introduced task priorities in ATSISBot.

- Update assertions in StorageTest.java to check for priority in task output
- Ensure tests reflect the new priority feature in task representations

Commit:
Merge 'edit-storage-junittest' into 'master'

Incorporate the updated JUnit tests from 'edit-storage-junittest' into the 'master' branch. These tests have been revised to match the new functionality related to task priority levels within the ATSISBot application.

- Align StorageTest.java with recent priority feature updates
- Ensure all unit tests are consistent with current output expectations

Commit:
Add missing JavaDocs to improve code documentation

Enhanced source code documentation by adding missing JavaDoc comments across various classes and methods. This provides better clarity and understanding of the codebase, facilitates easier maintenance, and aids new developer onboarding.

- Complete JavaDocs for previously undocumented public classes and methods
- Ensure consistency and completeness in documentation across the codebase

Commit:
Merge 'branch-A-missing-javadoc' into 'master'

Merge the comprehensive JavaDoc updates from 'branch-A-missing-javadoc' into the 'master' branch. This update fills in all previously missing JavaDoc comments, enhancing the overall documentation and readability of the code.

- Integrate full JavaDoc documentation into the master branch
- Improve code maintainability and developer understanding
